### ISSUE TITLE
Add new tool function for mocking agents in wazuh-manager

### ISSUE THREAD

**Issue by jmv74211** [wazuh/wazuh-qa#1049]
Hi,

To improve the wazuh testing, it is requested to create a new tool function for mocking agents in wazuh manager.

So far, we have an agent simulator, but in this case, we need to do something simpler, like making the manager believe that it has some agents already registered (even though they are not reporting anything)

For example, this new utility would be useful for vuln detector tests, since we can mock the agent system data instead of manager, avoiding possible conflicts when restarting `wazuh-modulesd` daemon, since after restarting it, the mocking data of `global.db` is lost.

To sum up, it is required to investigate how we can mock this agent, creating new data tables...

Regards.

**Comment by SpotHound** [comment]
We're working on this.

**Comment by SpotHound** [comment]
We need to add a new entry on the agent table at the global db for the new agent:

This is the structure of the table:
```
sqlite> PRAGMA table_info(agent);
0|id|INTEGER|0||1
1|name|TEXT|1||0
2|ip|TEXT|0||0
3|register_ip|TEXT|0||0
4|internal_key|TEXT|0||0
5|os_name|TEXT|0||0
6|os_version|TEXT|0||0
7|os_major|TEXT|0||0
8|os_minor|TEXT|0||0
9|os_codename|TEXT|0||0
10|os_build|TEXT|0||0
11|os_platform|TEXT|0||0
12|os_uname|TEXT|0||0
13|os_arch|TEXT|0||0
14|version|TEXT|0||0
15|config_sum|TEXT|0||0
16|merged_sum|TEXT|0||0
17|manager_host|TEXT|0||0
18|node_name|TEXT|0|'unknown'|0
19|date_add|INTEGER|1||0
20|last_keepalive|INTEGER|0||0
21|group|TEXT|0|'default'|0
22|sync_status|TEXT|1|'synced'|0
23|connection_status|TEXT|1|'never_connected'|0
```

When we try to manually add an agent we got: `Error: CHECK constraint failed: agent`

It seems that there is a constraint on the table. We check the table definition:

```
sqlite> select sql from sqlite_master where tbl_name = 'agent';
CREATE TABLE agent (    id INTEGER PRIMARY KEY,    name TEXT NOT NULL,    ip TEXT,    register_ip TEXT,    internal_key TEXT,    os_name TEXT,    os_version TEXT,    os_major TEXT,    os_minor TEXT,    os_codename TEXT,    os_build TEXT,    os_platform TEXT,    os_uname TEXT,    os_arch TEXT,    version TEXT,    config_sum TEXT,    merged_sum TEXT,    manager_host TEXT,    node_name TEXT DEFAULT 'unknown',    date_add INTEGER NOT NULL,    last_keepalive INTEGER,    `group` TEXT DEFAULT 'default',    sync_status TEXT NOT NULL CHECK (sync_status IN ('synced', 'syncreq')) DEFAULT 'synced',    connection_status TEXT NOT NULL CHECK (connection_status IN ('pending', 'never_connected', 'active', 'disconnected')) DEFAULT 'never_connected')
CREATE INDEX agent_name ON agent (name)
CREATE INDEX agent_ip ON agent (ip)
```

Some interesting constrains:

- agent id must be a number (1 instead of 001)
- agent status must be one of "connected", "disconnected", "never_connected"
- sync_status must be one of "synced" or "syncreq"

**Comment by SpotHound** [comment]
An example of valid query to add an agent to the db:

```
INSERT INTO AGENT
           ("id", "name", "ip", "register_ip", "internal_key", "os_name", "os_version", "os_major", "os_minor",
           "os_codename", "os_build", "os_platform", "os_uname", "os_arch", "version", "config_sum", "merged_sum",
           "manager_host", "node_name", "date_add", "last_keepalive", "group", "sync_status", "connection_status")
           VALUES
           (1, "centos8-agent", "127.0.0.1", "127.0.0.1", "", "CentOS Linuxs_name", "7.1", "7", "1",
           "centos-8", "4.18.0-147.8.1.el8_1.x86_64 ", " #1 SMP Thu Apr 9 13:49:54 UTC 2020", "x86_64", "x86_64", "4.2", "", "",
           "centos-8", "node01", "1612942494", "253402300799", "", "synced", "active")
```

**Comment by SpotHound** [comment]
I've created the following draft method to add new agents to the global db:

```
def create_agent(id=None, name="centos8-agent", ip="127.0.0.1", register_ip="127.0.0.1", internal_key="",
                 os_name="CentOS Linuxs_name", os_version="7.1", os_major="7", os_minor="1", os_codename="centos-8",
                 os_build="4.18.0-147.8.1.el8_1.x86_64", os_platform="#1 SMP Thu Apr 9 13:49:54 UTC 2020",
                 os_uname="x86_64", os_arch="x86_64", version="4.2", config_sum="", merged_sum="",
                 manager_host="centos-8", node_name="node01", date_add="1612942494", last_keepalive="253402300799",
                 group="", sync_status="synced", connection_status="active"):

    if id is None:
        query_last_id = 'select id from agent order by id desc limit 1;'
        last_id = get_query_result(GLOBAL_DB_PATH, query_last_id)
        id = int(last_id[0][0]) + 1

    query_create_agent = f'''INSERT INTO AGENT
           (id, name, ip, register_ip, internal_key, os_name, os_version, os_major, os_minor, 
            os_codename, os_build, os_platform, os_uname, os_arch, version, config_sum, merged_sum, 
            manager_host, node_name, date_add, last_keepalive, "group", sync_status, connection_status)
           VALUES
           ( "{id}", "{name}", "{ip}", "{register_ip}", "{internal_key}", "{os_name}", "{os_version}",
             "{os_major}", "{os_minor}", "{os_codename}", "{os_build}", "{os_platform}", "{os_uname}",
             "{os_arch}", "{version}", "{config_sum}", "{merged_sum}", "{manager_host}", "{node_name}",
             "{date_add}", "{last_keepalive}", "{group}", "{sync_status}", "{connection_status}")
           '''

    try:
        make_query(GLOBAL_DB_PATH, [query_create_agent])
    except sqlite3.IntegrityError:
        print("Could not add agent. Check if the selected id is already used by another agent.")
```

**Comment by SpotHound** [comment]
For some reason, after restarting modulesd, the globaldb is modified and the new entries are removed.

**Comment by SpotHound** [comment]
We've added some more changes:

1. We need to copy the 000.db entry from manager for each added agent (001.db, 002.db... etc)
2. We need to add an entry for the new agent in the client.keys file


```
def update_client_keys_entry(agent_id, agent_name, agent_ip="any", agent_key="31a4281e09b71809c7f7818ba27a0679462f139261cd086d21b80eef75ce020f"):
    with open("/var/ossec/etc/client.keys", 'r') as client_keys:
        client_entries = client_keys.readlines()

    with open("/tmp/new_client.keys", 'w') as tmp_keys:
        for entry in client_entries:
            _agent_id, _agent_name, _agent_ip, _agent_key = entry.split()
            if agent_id != _agent_id:
                tmp_keys.write(f"{_agent_id} {_agent_name} {_agent_ip} {_agent_key}\n")

        tmp_keys.write(f"{agent_id} {agent_name} {agent_ip} {agent_key}\n")

    with open("/var/ossec/etc/client.keys", 'w') as client_keys, open("/tmp/new_client.keys", 'r') as tmp_keys:
        client_keys.write(tmp_keys.read())

def create_agent(id=None, name="centos8-agent", ip="127.0.0.1", register_ip="127.0.0.1", internal_key="",
                 os_name="CentOS Linux", os_version="7.1", os_major="7", os_minor="1", os_codename="centos-8",
                 os_build="4.18.0-147.8.1.el8_1.x86_64", os_platform="#1 SMP Thu Apr 9 13:49:54 UTC 2020",
                 os_uname="x86_64", os_arch="x86_64", version="4.2", config_sum="", merged_sum="",
                 manager_host="centos-8", node_name="node01", date_add="1612942494", last_keepalive="253402300799",
                 group="", sync_status="synced", connection_status="active"):

    if id is None:
        query_last_id = 'select id from agent order by id desc limit 1;'
        last_id = get_query_result(GLOBAL_DB_PATH, query_last_id)
        id = int(last_id[0][0]) + 1

    query_create_agent = f'''INSERT INTO AGENT
           (id, name, ip, register_ip, internal_key, os_name, os_version, os_major, os_minor, 
            os_codename, os_build, os_platform, os_uname, os_arch, version, config_sum, merged_sum, 
            manager_host, node_name, date_add, last_keepalive, "group", sync_status, connection_status)
           VALUES
           ( "{id}", "{name}", "{ip}", "{register_ip}", "{internal_key}", "{os_name}", "{os_version}",
             "{os_major}", "{os_minor}", "{os_codename}", "{os_build}", "{os_platform}", "{os_uname}",
             "{os_arch}", "{version}", "{config_sum}", "{merged_sum}", "{manager_host}", "{node_name}",
             "{date_add}", "{last_keepalive}", "{group}", "{sync_status}", "{connection_status}")
           '''

    id_str = str(id).zfill(3)

    update_client_keys_entry(id_str, name, agent_ip=ip)
    os.cp("/var/ossec/queue/db/000.db", f"/var/ossec/queue/db/{id_str}.db")

    try:
        make_query(GLOBAL_DB_PATH, [query_create_agent])
    except sqlite3.IntegrityError:
        print("Could not add agent. Check if the selected id is already used by another agent.")
```

**Comment by SpotHound** [comment]
# Testing

## Alerts generated for this agent 
![image](https://user-images.githubusercontent.com/15269938/107518528-67ceff80-6baf-11eb-9681-6ec104eb25b6.png)
![image](https://user-images.githubusercontent.com/15269938/107518544-6d2c4a00-6baf-11eb-98b5-9d53c5c4ce13.png)

## Addig agents
![image](https://user-images.githubusercontent.com/15269938/107519149-31de4b00-6bb0-11eb-85d4-818127ed3292.png)
![image](https://user-images.githubusercontent.com/15269938/107519160-34d93b80-6bb0-11eb-9ccc-e68c1d6a57f0.png)
![image](https://user-images.githubusercontent.com/15269938/107519203-43275780-6bb0-11eb-9a77-48d2effbb9f5.png)

**Comment by SpotHound** [comment]
After some testing I've added the code to this commit: https://github.com/wazuh/wazuh-qa/commit/0e8271ccd1a175009c499b4f951593ed23729de8

on branch [1049-mocking-agent](https://github.com/wazuh/wazuh-qa/commits/1049-mocking-agent)
